[FG:InPlacePodVerticalScaling] Add extended resources to ContainerStatuses[i].Resources

iholder101 opened the pull request:
#### What type of PR is this?
/kind feature
/sig node

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
In-place pod resize feature adds Resources field to ContainerStatuses struct. This is used to report CPU and memory requests & limits configured for the containers.

After this PR, extended resources would also be reflected as part of `ContainerStatuses[i].Resources`.

#### Sample output:
Add a dummy extended resource to the node:
```bash
> k describe node | grep Allocatable -w -A5
Allocatable:
  cpu:                40
  dummy.com/dummy:    4  # <---- Extended resource
  ephemeral-storage:  1293695812Ki
  hugepages-1Gi:      0
  hugepages-2Mi:      0
```

Add a pod using this resource:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: dummy-pod
spec:
  containers:
  - name: dummy-container
    image: k8s.gcr.io/pause:2.0
    resources:
      requests:
        cpu: 100m
        dummy.com/dummy: 1
      limits:
        dummy.com/dummy: 1
```

Create the pod and check its `ContainerStatuses[i].Resources`:
```bash
> k get pod -o yaml | grep containerstatus -i -A40
    containerStatuses:
    - allocatedResources:
        cpu: 100m
        dummy.com/dummy: "1"
      # ....
      resources:
        limits:
          dummy.com/dummy: "1"
        requests:
          cpu: 100m
          dummy.com/dummy: "1"
```

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #114159

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Added status for extended Pod resources within the `status.containerStatuses[].resources` field.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources
```

iholder101 committed:
Add extended resources to ContainerStatuses[i].Resources

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
unit test: Add extended resources to ContainerStatuses[i].Resources

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
Refactor: avoid code duplication when converting custom resources

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
Remove TODO regarding extended resources

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
/test pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2

iholder101 commented:
/cc @vinaykul

iholder101 commented:
riage accepted
/priority important-longterm

iholder101 commented:
/triage accepted

sftim commented:
Changelog suggestion

```diff
-Extended resources to be reflected as part of the pod.Status.ContainerStatuses[i].Resources struct.
+Added status for extended Pod resources within the `status.containerStatuses[].resources` field.
```

sftim commented:
Changelog suggestion

```diff
-Extended resources to be reflected as part of the pod.Status.ContainerStatuses[i].Resources struct.
+Added status for extended Pod resources within the `status.containerStatuses[].resources` field.
```

vinaykul commented:
@iholder101 Thanks for the PR! 😄 

If it is not too much to ask, can you please extend the E2E tests test/e2e/node/pod_resize.go:doPodResizeTests() by adding a test case that requests and verifies extended resources? 

Eyeballing the code, verifyPodAllocations() and verifyPodStatusResources() should already check for it but please confirm that it indeed verifies extended resources as well. I think the nodes need to be patched for extended resources .. a possible example for reference: https://github.com/kubernetes/kubernetes/blob/9791f0d1f39f3f1e0796add7833c1059325d5098/test/e2e/scheduling/preemption.go#L139-L144

vinaykul commented:
Also adding a few more sharp eyes for this PR.
/cc @mrunalp @bobbypage @derekwaynecarr @dchen1107

vinaykul commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -2071,6 +2071,18 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
 		}
+
+		convertCustomResources := func(inResources, outResources v1.ResourceList) {
+			for extendedResourceName, extendedResourceQuantity := range inResources {
```
resourceName, resourceQuantity sounds more appropriate instead of extendedResourceName, extendedResourceQuantity

vinaykul commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -2071,6 +2071,18 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
+					extendedResourceName == v1.ResourceStorage || extendedResourceName == v1.ResourceEphemeralStorage {
+					continue
+				}
+
```
nit: remove this blank line please

vinaykul commented on pkg/kubelet/kubelet_pods_test.go:
```diff
@@ -4737,6 +4745,52 @@ func TestConvertToAPIContainerStatusesForResources(t *testing.T) {
+				},
+			},
+		},
 	} {
```
Perhaps a test case with only ExendedResources?

iholder101 replied:
Isn't `"BestEffort QoSPod with extended resources"` covering that already? Can you please elaborate?

vinaykul reviewed (commented):
A few minor comments..

iholder101 committed:
Refactor: improve convertCustomResources() variable names

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
> @iholder101 Thanks for the PR! 😄
> 
> If it is not too much to ask, can you please extend the E2E tests test/e2e/node/pod_resize.go:doPodResizeTests() by adding a test case that requests and verifies extended resources?
> 
> Eyeballing the code, verifyPodAllocations() and verifyPodStatusResources() should already check for it but please confirm that it indeed verifies extended resources as well. I think the nodes need to be patched for extended resources .. a possible example for reference:
> 
> https://github.com/kubernetes/kubernetes/blob/9791f0d1f39f3f1e0796add7833c1059325d5098/test/e2e/scheduling/preemption.go#L139-L144

You're welcome :)

Thanks a lot for your detailed guidance and pointers.
I've added an e2e test with extended resources. I also believe I've addressed the other comments you've posted.

Can you PTAL?

iholder101 commented:
/test pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2

vinaykul commented:
> > @iholder101 Thanks for the PR! 😄
> > If it is not too much to ask, can you please extend the E2E tests test/e2e/node/pod_resize.go:doPodResizeTests() by adding a test case that requests and verifies extended resources?
> > Eyeballing the code, verifyPodAllocations() and verifyPodStatusResources() should already check for it but please confirm that it indeed verifies extended resources as well. I think the nodes need to be patched for extended resources .. a possible example for reference:
> > https://github.com/kubernetes/kubernetes/blob/9791f0d1f39f3f1e0796add7833c1059325d5098/test/e2e/scheduling/preemption.go#L139-L144
> 
> You're welcome :)
> 
> Thanks a lot for your detailed guidance and pointers. I've added an e2e test with extended resources. I also believe I've addressed the other comments you've posted.
> 
> Can you PTAL?

I'll try and get to this over the coming weekend. But the CI job has failed, please look.
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/124227/pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2/1781007162655379456

iholder101 committed:
Add an e2e test

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
/test pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2

vinaykul commented:
/test pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2

vinaykul commented:
/test

vinaykul commented:
/test pull-kubernetes-e2e-gce-cos-alpha-features

vinaykul reviewed (approved):
@iholder101 Thanks a lot for this work. I tried out the patch as well and it looks good. I have one suggestion: add the extended resource to all nodes in the cluster and let scheduler handle pod binding. Please review my comments.
After that change, for me this is 
/lgtm

vinaykul commented:
@iholder101 The below incremental change to your PR should do the trick!
```bash
root@vbuild:~/go/src/k8s.io/kubernetes# git diff
diff --git a/test/e2e/node/pod_resize.go b/test/e2e/node/pod_resize.go
index 91640435758..83918195d42 100644
--- a/test/e2e/node/pod_resize.go
+++ b/test/e2e/node/pod_resize.go
@@ -1347,7 +1347,6 @@ func doPodResizeTests() {
                ginkgo.It(tc.name, func(ctx context.Context) {
                        var testPod, patchedPod *v1.Pod
                        var pErr error
-                       var nodeName string
 
                        tStamp := strconv.Itoa(time.Now().Nanosecond())
                        initDefaultResizePolicy(tc.containers)
@@ -1358,10 +1357,14 @@ func doPodResizeTests() {
                                nodes, err := e2enode.GetReadySchedulableNodes(context.Background(), f.ClientSet)
                                framework.ExpectNoError(err)
 
-                               nodeName = nodes.Items[0].Name
-
-                               addExtendedResource(f.ClientSet, nodeName, fakeExtendedResource, resource.MustParse("123"))
-                               testPod.Spec.NodeName = nodeName
+                               for _, node := range nodes.Items {
+                                       addExtendedResource(f.ClientSet, node.Name, fakeExtendedResource, resource.MustParse("123"))
+                               }
+                               defer func() {
+                                       for _, node := range nodes.Items {
+                                               removeExtendedResource(f.ClientSet, node.Name, fakeExtendedResource)
+                                       }
+                               }()
                        }
 
                        ginkgo.By("creating pod")
@@ -1416,10 +1419,6 @@ func doPodResizeTests() {
                        ginkgo.By("deleting pod")
                        err = e2epod.DeletePodWithWait(ctx, f.ClientSet, newPod)
                        framework.ExpectNoError(err, "failed to delete pod")
-
-                       if tc.addExtendedResource {
-                               removeExtendedResource(f.ClientSet, nodeName, fakeExtendedResource)
-                       }
                })
        }
 }
```

iholder101 committed:
Patch all nodes with extended resource to allow testing on every node

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
Thanks a lot @vinaykul! Awesome suggestion!
Done :) PTAL

vinaykul commented:
/test pull-kubernetes-e2e-gce-cos-alpha-features
/test pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2

vinaykul reviewed (approved):
/lgtm

vinaykul commented:
/test pull-kubernetes-e2e-gce-cos-alpha-features
/test pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2

SergeyKanzhelev commented:
/assign @tallclair

SergeyKanzhelev commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -2072,6 +2072,17 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
+		convertCustomResources := func(inResources, outResources v1.ResourceList) {
+			for resourceName, resourceQuantity := range inResources {
+				if resourceName == v1.ResourceCPU || resourceName == v1.ResourceMemory ||
+					resourceName == v1.ResourceStorage || resourceName == v1.ResourceEphemeralStorage {
```
why `resourceName == v1.ResourceStorage` excluded here and have no special processing below? Is it intended? Please also add a test on it

iholder101 replied:
I'm not sure I understand the question.
When `resourceName == v1.ResourceStorage` the resource is not a custom resource, and it's already being handled below:
https://github.com/kubernetes/kubernetes/blob/a31030543c47aac36cf323b885cfb6d8b0a2435f/pkg/kubelet/kubelet_pods.go#L2095-L2096
https://github.com/kubernetes/kubernetes/blob/a31030543c47aac36cf323b885cfb6d8b0a2435f/pkg/kubelet/kubelet_pods.go#L2112-L2113

What am I missing?

tallclair replied:
I see special handling for ephemeral storage, but not `v1.ResourceStorage`. Moreover, the ephemeral storage conversion is doing the same thing that's being done here, so it might as well be combined with this conversion.

iholder101 replied:
Thanks for the clarifications! :)
Done! Also added a unit test.

tallclair commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -2072,6 +2072,17 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
+					resourceName == v1.ResourceStorage || resourceName == v1.ResourceEphemeralStorage {
+					continue
+				}
+				outResources[resourceName] = resourceQuantity
```
should this be a `.DeepCopy()`?

vinaykul replied:
Good catch. It should be DeepCopy for consistency.

iholder101 replied:
Done! Thanks for catching that!

tallclair commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -2088,6 +2099,8 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
 				limits[v1.ResourceEphemeralStorage] = ephemeralStorage.DeepCopy()
 			}
+
+			convertCustomResources(container.Resources.Limits, limits)
```
Alternatively: just deep-copy the list, and then special-case the memory & cpu limits. E.g.

```go
container.Resources.Limits = limits.DeepCopy()
if cStatus.Resources != nil && cStatus.Resources.CPULimit != nil {
  // ...
```

tallclair commented on test/e2e/node/pod_resize.go:
```diff
@@ -18,7 +18,9 @@ package node
 	"context"
+	"encoding/json"
 	"fmt"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
```
nit: move to k8s import block below

iholder101 replied:
Done! :)

tallclair commented:
Who is the intended consumer for these resources? I think the status is supposed to reflect the actual state of resources in-use, so it seems like if we're including extended resources there should be some sort of confirmation from device drivers that the resources are in fact in use.

Also, AFAICT this isn't including extended resources in the allocated resources. Should it?

vinaykul commented:
> Who is the intended consumer for these resources? I think the status is supposed to reflect the actual state of resources in-use, so it seems like if we're including extended resources there should be some sort of confirmation from device drivers that the resources are in fact in use.
> 
> Also, AFAICT this isn't including extended resources in the allocated resources. Should it?

@tallclair AFAIK extended resources are not managed by the CRI. It is used for scheduling and kubelet is admission gatekeeper. If a pod requesting extended resources is admitted, it can be assumed that the pod's containers got the extended resources they requested. And since it is immutable (at this time), reflecting it back in status just like ephemeral-storage would be accurate.

The AllocatedResources part is a bit non-obvious. It gets checkpointed upon admission and resize: https://github.com/kubernetes/kubernetes/blob/4bb434501d9ee5edda6faf52a9d6d32a969ae183/pkg/kubelet/kubelet.go#L2576-L2592 
And it is retrieved from checkpoint during APIStatus generation:
https://github.com/kubernetes/kubernetes/blob/4bb434501d9ee5edda6faf52a9d6d32a969ae183/pkg/kubelet/kubelet_pods.go#L2074-L2083 

I suppose you can see the fragility of this code! Some added context: node-local checkpointing was much debated topic during design discussions.. Dawn and I eventually capitulated in order to break the impasse and make progress. But at that time we didn’t have KEP https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/2527-clarify-status-observations-vs-rbac and the precedent PR https://github.com/kubernetes/kubernetes/pull/119665. I feel we can drop all of the checkpointing code and rely on pod.Status but the APF aspect needs a closer look.

iholder101 committed:
Add limits and request processing for storage resources

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
unit test: Add limits and request processing for storage resources

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
@tallclair @vinaykul @SergeyKanzhelev 

Sorry for the delay. I was out of office for a while.
Would you mind having another look?

tallclair commented:
/lgtm
/approve
